### PR TITLE
feat: re-structure `basecoin-rs`

### PR DESCRIPTION
**Pull request by Farhad-Shabani**
Closes: #155

### Along with this PR
- Incorporate `rustfmt.toml` to leverage the Rust nightly formatter for easier imports' organization.
- Update Hermes to `v1.8.0`

### CODE REVIEW

File: basecoin/modules/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+repository  = { workspace = true }
+authors     = { workspace = true }
+description = """
+        Contains simple implementations of various modules for the Basecoin application.
+"""
+
+[dependencies]
+# external dependencies
+base64             = { workspace = true }
+cosmrs             = "0.15.0"
```
Comment (rnbguy) [inline basecoin/modules/Cargo.toml:16]:
```suggestion
cosmrs             = { version = "0.15.0" }
```

Comment (rnbguy) [reply basecoin/modules/Cargo.toml:16]: should we make this workspace dependency?

File: basecoin/app/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,41 @@
+[package]
+name       = "basecoin-app"
+version     = { workspace = true }
+edition     = { workspace = true }
+license     = { workspace = true }
+readme      = { workspace = true }
+repository  = { workspace = true }
+authors     = { workspace = true }
+description  = """
+    Contains a rudimentary implementation of the Tendermint ABCI application.
```
Comment (rnbguy) [inline basecoin/app/Cargo.toml:10]:
```suggestion
    Contains a simple implementation of the Tendermint ABCI application.
```

Just a suggestion. You can ignore this.

File: basecoin/src/helper.rs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+use basecoin_modules::auth::{AuthAccountKeeper, AuthAccountReader};
+use basecoin_modules::bank::Bank;
+use basecoin_modules::context::{prefix, Identifiable};
+use basecoin_modules::ibc::Ibc;
+use basecoin_store::context::ProvableStore;
+use basecoin_store::impls::RevertibleStore;
+use basecoin_store::utils::SharedRwExt;
+
+/// Gives access to the IBC module.
+pub fn ibc<S>(app: BaseCoinApp<S>) -> Ibc<RevertibleStore<S>>
```
Comment (rnbguy) [inline basecoin/src/helper.rs:13]: we never used `ibc` and `bank` functions. can we remove this file ?

**Review by rnbguy — CHANGES_REQUESTED** [review]
Thanks @Farhad-Shabani for this PR ! :clap:

Very helpful to maintain cross-project dependencies.

**Comment by rnbguy** [comment]
The integration test fails because the client upgrade case has failed.

File: README.md [hunk 4]
```diff
@@ -49,20 +49,19 @@ $ cargo run -- --help
 basecoin 0.1.0
 
 USAGE:
-    basecoin [FLAGS] [OPTIONS]
+    main [FLAGS] [OPTIONS]
```
[the remark this answers is not in the document]
Comment (Farhad-Shabani) [reply README.md]: [ed23af3](https://github.com/informalsystems/basecoin-rs/pull/156/commits/ed23af315bc0755b0de03f466ae89a9f5b6d6a31)

File: basecoin/app/src/abci/v0_38/tendermint.rs [hunk 5]
```diff
@@ -56,7 +49,7 @@ impl<S: Debug + ProvableStore> Application for BaseCoinApp<S> {
 
         let mut modules = self.modules.write_access();
 
-        for IdentifiedModule { module, id } in modules.iter_mut() {
+        for IdentifiedModule { module, id: _ } in modules.iter_mut() {
```
[the remark this answers is not in the document]
Comment (Farhad-Shabani) [reply basecoin/app/src/abci/v0_38/tendermint.rs]: [ed23af3](https://github.com/informalsystems/basecoin-rs/pull/156/commits/ed23af315bc0755b0de03f466ae89a9f5b6d6a31)

File: basecoin/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+[dependencies]
+# external dependencies
+clap               = { version = "4.3.0", features = ["derive", "wrap_help"] }
+ibc-proto          = { workspace = true, features = ["server", "proto-descriptor", "serde"] }
+serde              = { workspace = true }
+serde_derive       = { version = "1.0.104", default-features = false }
+tokio              = { version = "1.0", features = ["macros", "rt-multi-thread"] }
+tonic              = { workspace = true }
+tonic-reflection   = { workspace = true }
+toml               = "0.8"
```
[the remark this answers is not in the document]
Comment (Farhad-Shabani) [reply basecoin/Cargo.toml]: [ed23af3](https://github.com/informalsystems/basecoin-rs/pull/156/commits/ed23af315bc0755b0de03f466ae89a9f5b6d6a31)

File: basecoin/modules/Cargo.toml (hunk 1 above)
Comment (Farhad-Shabani) [reply basecoin/modules/Cargo.toml]: [ed23af3](https://github.com/informalsystems/basecoin-rs/pull/156/commits/ed23af315bc0755b0de03f466ae89a9f5b6d6a31)

File: basecoin/app/Cargo.toml (hunk 2 above)
Comment (Farhad-Shabani) [reply basecoin/app/Cargo.toml]: [ed23af3](https://github.com/informalsystems/basecoin-rs/pull/156/commits/ed23af315bc0755b0de03f466ae89a9f5b6d6a31)

File: basecoin/src/helper.rs (hunk 3 above)
Comment (Farhad-Shabani) [reply basecoin/src/helper.rs:13]: In downstream projects, these serve as convenient methods callable on an instance of `BaseCoinApp<S>`, providing access to either ibc or bank modules. Let's retain these and observe how changes would appear in the `ibc-testkit`. If unnecessary, let's remove them at that time.

File: basecoin/modules/src/upgrade/query.rs [hunk 7]
```diff
@@ -1,16 +1,15 @@
...
 use super::path::UpgradePlanPath;
-use crate::types::config::CometbftConfig;
-use crate::types::error::Error;
+use crate::error::Error;
 
 pub(crate) const UPGRADE_PLAN_QUERY_PATH: &str = "/cosmos.upgrade.v1beta1.Query/CurrentPlan";
 
-pub async fn query_upgrade_plan(cfg: CometbftConfig) -> Result<Plan, Error> {
-    let rpc_client = HttpClient::new(cfg.rpc_addr.clone()).unwrap();
+pub async fn query_upgrade_plan(rpc_addr: Url) -> Result<Plan, Error> {
```
[the remark this answers is not in the document]
Comment (Farhad-Shabani) [reply basecoin/modules/src/upgrade/query.rs:11]: That's right. Specifically, as the `config.rs` will be living in the `basecoin` crate, not accessible here.

File: ci/hermes-config.toml [hunk 8]
```diff
@@ -74,8 +85,6 @@ proof_specs = '''
   }
 ]
 '''
+trust_threshold = '2/3'
```
[the remark this answers is not in the document]
Comment (Farhad-Shabani) [reply ci/hermes-config.toml:88]: Good question. Revised for consistency with the latest Hermes v1.8 configuration.

File: README.md [hunk 9]
```diff
@@ -49,20 +49,19 @@ $ cargo run -- --help
...
-    -g, --grpc-port <grpc-port>            Bind the gRPC server to this port [default: 9093]
-    -h, --host <host>                      Bind the TCP server to this host [default: 127.0.0.1]
-    -p, --port <port>                      Bind the TCP server to this port [default: 26658]
-    -r, --read-buf-size <read-buf-size>    The default server read buffer size, in bytes, for each incoming client
-                                           connection [default: 1048576]
+        --config <FILE>  The path to the configuration file [default: config.toml]
+    -v, --verbose        Increase output logging verbosity to DEBUG level
+    -q, --quiet          Suppress all output logging (overrides --verbose)
+    -h  --help           Prints help information
+    -V, --version        Prints version information
```
[the remark this answers is not in the document]
Comment (Farhad-Shabani) [reply README.md]: [ed23af3](https://github.com/informalsystems/basecoin-rs/pull/156/commits/ed23af315bc0755b0de03f466ae89a9f5b6d6a31)

File: .changelog/unreleased/breaking-changes/156-re-structure-repo.md [hunk 10]
```diff
@@ -0,0 +1,2 @@
+- Re-structure repository with the focus on `ibc-testkit` integration
```
[the remark this answers is not in the document]
Comment (Farhad-Shabani) [reply .changelog/unreleased/breaking-changes/156-re-structure-repo.md]: Oops. [ed23af3](https://github.com/informalsystems/basecoin-rs/pull/156/commits/ed23af315bc0755b0de03f466ae89a9f5b6d6a31)

File: basecoin/app/src/service.rs [hunk 11]
```diff
@@ -66,12 +46,19 @@ impl<S: ProvableStore> HealthService for BaseCoinApp<S> {
...
+                build_deps: vec![
+                    VersionInfoModule {
+                        path: "github.com/cometbft/cometbft".to_string(),
+                        version: "v0.37.1".to_string(),
+                        sum: "".to_string(),
+                    },
+                    VersionInfoModule {
+                        path: "github.com/cosmos/cosmos-sdk".to_string(),
+                        version: "v0.47.0".to_string(),
+                        sum: "h1:ps1QWfvaX6VLNcykA7wzfii/5IwBfYgTIik6NOVDq/c=".to_string(),
```
[the remark this answers is not in the document]
Comment (Farhad-Shabani) [reply basecoin/app/src/service.rs]: Updated: [f4bc959](https://github.com/informalsystems/basecoin-rs/pull/156/commits/f4bc9595f79e0cc94e04fa5de6ff3ee5f0bbdc4d)

File: basecoin/app/src/service.rs [hunk 12]
```diff
@@ -66,12 +46,19 @@ impl<S: ProvableStore> HealthService for BaseCoinApp<S> {
...
-                    path: "github.com/cosmos/cosmos-sdk".to_string(),
-                    version: "v0.47.0".to_string(),
-                    sum: "h1:ps1QWfvaX6VLNcykA7wzfii/5IwBfYgTIik6NOVDq/c=".to_string(),
-                }],
+                git_commit: "44bae428392201d541cc2811de4369ea664c5762".to_string(),
+                build_deps: vec![
+                    VersionInfoModule {
+                        path: "github.com/cometbft/cometbft".to_string(),
+                        version: "v0.37.1".to_string(),
+                        sum: "".to_string(),
```
[the remark this answers is not in the document]
Comment (Farhad-Shabani) [reply basecoin/app/src/service.rs]: [f4bc959](https://github.com/informalsystems/basecoin-rs/pull/156/commits/f4bc9595f79e0cc94e04fa5de6ff3ee5f0bbdc4d)

File: basecoin/src/bin/basecoin.rs [hunk 13]
```diff
@@ -36,10 +35,10 @@ async fn main() {
         Commands::Query(q) => {
             let query_res = match q {
                 QueryCmd::Upgrade(u) => match u {
-                    UpgradeCmd::Plan => query_upgrade_plan(cfg.cometbft).await.unwrap(),
+                    UpgradeCmd::Plan => query_upgrade_plan(cfg.cometbft.rpc_addr).await.unwrap(),
                 },
             };
-            println!("{:?}", query_res);
+            let _ = std::io::stdout().write(format!("{:#?}", query_res).as_bytes());
```
Comment (rnbguy) [inline basecoin/src/bin/basecoin.rs:41]: More succinct.

```suggestion
            let _ = write!(std::io::stdout(), "{:#?}", query_res);
```

File: .github/workflows/integration.yml [hunk 14]
```diff
@@ -12,14 +12,14 @@ jobs:
...
         env:
           GITHUB_TOKEN: "${{ secrets.GITHUB_TOKEN }}"
-    if: "!startsWith(github.ref, 'refs/tags/') && github.ref != 'refs/heads/master'"
+    if: ${{ !startsWith(github.ref, 'refs/tags/') && github.ref != 'refs/heads/master' }}
 
   ibc-integration:
     runs-on: ubuntu-latest
     env:
       COMETBFT_VERSION: 0.37.1
       GAIA_VERSION: 6.0.0
```
Comment (rnbguy) [inline .github/workflows/integration.yml:21]: Dockerfile and workflow files should be in sync.

```suggestion
      GAIA_VERSION: 7.1.1
```

File: basecoin/app/src/service.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (rnbguy) [reply basecoin/app/src/service.rs]: awesome ! matches with my machine :slightly_smiling_face:

**Review by rnbguy — APPROVED** [review]
I added one last comment. After that, go ahead with the merge.

Also I updated the GAIA version on CI at 11bb6af

Great work ! :raised_hands:

File: basecoin/src/bin/basecoin.rs [hunk 15]
```diff
@@ -36,10 +35,10 @@ async fn main() {
         Commands::Query(q) => {
             let query_res = match q {
                 QueryCmd::Upgrade(u) => match u {
-                    UpgradeCmd::Plan => query_upgrade_plan(cfg.cometbft).await.unwrap(),
+                    UpgradeCmd::Plan => query_upgrade_plan(cfg.cometbft.rpc_addr).await.unwrap(),
                 },
             };
-            println!("{:?}", query_res);
+            let _ = write!(std::io::stdout(), "{:#?}", query_res);
```
[the remark this answers is not in the document]
Comment (Farhad-Shabani) [reply basecoin/src/bin/basecoin.rs:41]: I didn't perceive much distinction between these two methods here, so I kept this. We can consider it in a later PR, if you find the println! to be more effective?

**Pull request merged**